### DISCUSSION TITLE
404 status code when using option { output.filename: '[name].[contenthash].js' } in webpack.config.js

### DISCUSSION THREAD

**Discussion by OptimistLabyrinth** [webpack/webpack#13824] [General]
# Bug report

# **What is the current behavior?**

I apologize in advance that I'm not that good at English.

While I was following the webpack docs guide(https://webpack.js.org/guides/caching/), 
I came across an issue.
I'm not sure if this is really an issue or not, but I think it could be problematic.
I tried 3 cases to figure out when the issue occurs and when it doesn't

I used webpack-dev-server to host my webpack bundled app.

## package.json file

```json
{
  "name": "webpack-any",
  "version": "1.0.0",
  "description": "",
  "private": "true",
  "scripts": {
    "start": "npm run build && webpack server --config webpack.prod.js",
    "build": "webpack --config webpack.prod.js",
    "dev": "npm run devBuild && webpack server --config webpack.config.js",
    "devBuild" : "webpack --config webpack.config.js",
    "clean": "rimraf ./prod/*",
    "devClean": "rimraf ./dist/*"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@babel/core": "^7.14.6",
    "@babel/preset-env": "^7.14.7",
    "babel-loader": "^8.2.2",
    "css-loader": "^6.1.0",
    "csv-loader": "^3.0.3",
    "html-loader": "^2.1.2",
    "html-webpack-plugin": "^5.3.2",
    "rimraf": "^3.0.2",
    "style-loader": "^3.1.0",
    "webpack": "^5.45.1",
    "webpack-cli": "^4.7.2",
    "webpack-dev-server": "^3.11.2",
    "webpack-manifest-plugin": "^3.1.1"
  }
}

```

## 1. Setting output.filename = '[name].bundle.js' + output.publicPath = '/dist/

![name bundle js+publicPath=dist_OK](https://user-images.githubusercontent.com/16411622/126186963-67028286-2cd6-48dc-b0e6-3bba62568c02.png)

## 2. Setting output.filename = '[name].[contenthash].js' + output.publicPath = '/dist/

![name contenthash js+publicPath=dist_ERROR](https://user-images.githubusercontent.com/16411622/126187084-9e8fefe3-60ad-4d0f-9ddf-b976a78ca267.png)

## 3. Setting output.filename = '[name].[contenthash].js' + output.publicPath = 'auto'

![name contenthash js+publicPath=auto_OK](https://user-images.githubusercontent.com/16411622/126187094-90a0de45-5ac1-4e8d-b98e-1a6d362fc209.png)

## OK

# **If the current behavior is a bug, please provide the steps to reproduce.**

# **What is the expected behavior?**

So, I realized
when I use webpack caching machenism such as output.filename = [contenthash].js, I cannot use output.publicPath option.
This happens both in 'development' mode and 'production' mode.

Exprected behavior is
even though the webpack users set output.filename with [contenthash],
webpack output must be adjusted according to the output.publicPath value.

**Other relevant information:**
webpack version:  v5.45.1
Node.js version: v14.15.0
Operating System: ubuntu-20.04-LTS
Additional tools:

**Comment by alexander-akait** [comment]
Do not ignore this `If the current behavior is a bug, please provide the steps to reproduce.`

**Reply by OptimistLabyrinth** [reply]
I created a repository to easily reproduce the issue.
https://github.com/OptimistLabyrinth/webpack-discussion-13824

Hope it helps

**Reply by alexander-akait** [reply]
Why do you use the same directory for `contentBase` and webpack assets, `contentBase` is like `express-static`

**Reply by OptimistLabyrinth** [reply]
{
    devServer: {
        contentBase: ( A )
    }
}

are you referring to 'A' ??

**Reply by alexander-akait** [reply]
I don't understand you

**Reply by OptimistLabyrinth** [reply]
https://webpack.js.org/configuration/dev-server/#devservercontentbase

I think I've misunderstood devServer.contentBase.
according to docs, it is said 

```
Tell the server where to serve content from. This is only necessary if you want to serve static files. devServer.publicPath will be used to determine where the bundles should be served from, and takes precedence.
```

then I wonder how webpack-dev-server knows where the .html, .js, etc. files are which I send to when the HTTP client request comes in
Does it follow the output path?

**Reply by OptimistLabyrinth** [reply]
For my project, 
I build dev-mode into {root}/dist and for prod-mode into {root}/prod directory.

**Reply by OptimistLabyrinth** [reply]
Could you tell me the proper way to write webpack.config.js, webpack.prod.js ?
Maybe not the whole file but one or two corrections ?

**Reply by alexander-akait** [reply]
> Does it follow the output path?

Yes

**Reply by alexander-akait** [reply]
Also I don't see any problem with `publicPath: 'auto'`, all works fine

**Reply by OptimistLabyrinth** [reply]
yes it works fine but 
when I set { output.filename = [name].bundle.js + output.publicPath = 'somefolder' } my app works fine
however when I set { output.filename = [name].[contenthash].js + output.publicPath = 'somefolder' } my app always responds with 404 and I was forced to set { output.publicPath = 'auto' }.
currently I haven't released any product so there's no problem
I was being cautious in case I happen to be in a situation where I need to use both output.filename = [contenthash].js setting and output.publicPath setting.

#### 

is it troublesome to edit this part ?

**Reply by alexander-akait** [reply]
I don't have problem with `filename: '[name].[contenthash].js',` in your example

**Reply by OptimistLabyrinth** [reply]
that's strange

**Reply by OptimistLabyrinth** [reply]
I'l look for another factor that bring about it happens. I think it would take time.
Would you close this disscussion or just leave open ?

**Reply by alexander-akait** [reply]
Let's keep open, it is just dicussion
